src/main/java/net/neoforged/neoforge/network/registration/NetworkRegistry.java

```diff
@@ -0,0 +1,955 @@
+            return ModdedNetworkSetupFailedPayload.READER;
+        }
+
+        //Check the network setup.
+        final NetworkPayloadSetup payloadSetup = context.channel().attr(ATTRIBUTE_PAYLOAD_SETUP).get();
+        final PacketFlow flow = context.channel().attr(ATTRIBUTE_FLOW).get();
+        if (payloadSetup == null || flow == null) {
+            //Error: Bail.
+            LOGGER.warn("Received a modded custom payload packet {} that has not negotiated with the server. Not parsing packet.", id);
+            return null;
+        }
+
+        //Now ask the protocol what kind of payload is being sent and get the channel for it.
+        if (protocol.isPlay()) {
+            final NetworkChannel channel = payloadSetup.play().get(id);
+
+            //Validate that everything is okey and then return a reader.
+            if (channel == null) {
+                LOGGER.warn("Received a modded custom payload packet {} with an unknown or not accepted channel. Not parsing packet.", id);
+                return null;
+            }
+
+            final PlayRegistration<?> registration = knownPlayRegistrations.get(channel.id());
+            if (registration == null) {
+                LOGGER.error("Received a modded custom payload packet {} with an unknown or not accepted channel. Not parsing packet.", channel.id());
+                throw new IllegalStateException("A client send a packet with an unknown or not accepted channel, while negotiation succeeded. Somebody changed the channels known to NeoForge!");
+            }
+
+            if (registration.flow().isPresent()) {
+                if (registration.flow().get() != flow) {
+                    LOGGER.warn("Received a modded custom payload packet {} that is not supposed to be send to the server. Disconnecting client.", channel.id());
+                    final Connection connection = ConnectionUtils.getConnection(context);
+                    final PacketListener listener = connection.getPacketListener();
+                    if (listener instanceof ServerGamePacketListener serverListener) {
+                        serverListener.disconnect(Component.translatable("neoforge.network.invalid_flow"));
+                    } else if (listener instanceof ClientGamePacketListener clientListener) {
+                        clientListener.getConnection().disconnect(Component.translatable("neoforge.network.invalid_flow"));
+                    } else {
+                        LOGGER.error("Received a modded custom payload packet {} that is not supposed to be send to the server. Disconnecting client, but the listener is not a game listener. This should not happen.", channel.id());
+                        throw new IllegalStateException("A client send a packet with an unknown or not accepted channel, while negotiation succeeded. Somebody changed the channels known to NeoForge!");
+                    }
+                }
+            }
+
+            return registration;
```

Comment:
Can this block of duplicate code that is the say for if it is play or is configuration be merged somehow? Basically the unique parts of these two branches of the if statement are getting the:
- `NetworkChannel channel`
- `registration`

Granted we would need another internal interface to handle some of the hierarchy stuff between the two (such as for returning flow etc), but I think it might be worth it.

Reply:
Probably in some way, but it would become pretty ugly pretty quickly.
Unless somebody has a proposal, I suggest leaving it like this.